tests/e2e/tidbcluster/tidbcluster.go

```diff
@@ -1558,8 +1559,313 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
 		oa.UpgradeTidbClusterOrDie(&tcCfg)
 		oa.CheckTidbClusterStatusOrDie(&tcCfg)
 	})
+
+	// test cases for tc upgrade
+	ginkgo.Context("upgrade should work correctly", func() {
+		ginkgo.It("for tc and components version", func() {
+			ginkgo.By("Deploy initial tc")
+			tc := fixture.GetTidbCluster(ns, "upgrade-version", utilimage.TiDBV4Prev)
+			pvRetain := corev1.PersistentVolumeReclaimRetain
```

Comment:
Is it necessary to use `Retain` here?

Reply:
the detailed e2e test doc specifies "set PV reclaim policy to Retain"

Reply:
You're right, this case also covers the `Retain` and local PV scenarios, sorry that I forgot this.